thirdeye/thirdeye-frontend/app/pods/components/rootcause-trend/component.js

```diff
@@ -0,0 +1,389 @@
+        for (let i = 0; i < timeseries[baseUrn].timestamp.length; i++) {
+          baseValueLookup[timeseries[baseUrn].timestamp[i]] = timeseries[baseUrn].value[i];
+        }
+
+        timeseries[currUrn].timestamp.forEach((t, i) => {
+          const change = timeseries[currUrn].value[i] / (baseValueLookup[t] || Number.NaN) - 1;
+
+          // TODO use O(logN) data structure (e.g. treemap)
+          const index = buckets.findIndex(b => b >= t);
+          if (index >= 0 && index < buckets.length) {
+            changes[urn][index] = change;
+          }
+        });
+      });
+
+      console.log('changes done', changes);
+
+      return changes;
+    }
+  ),
+
+  /**
+   * Table data in rows
+   * @type {object[]}
+   */
+  data: computed(
+    'entities',
+    'buckets',
+    'changes',
+    'links',
+    'selectedUrns',
+    function () {
+      const { entities, buckets, changes, links, selectedUrns } =
+        this.getProperties('entities', 'buckets', 'changes', 'links', 'selectedUrns');
+
+      const metricUrns = filterPrefix(Object.keys(entities), 'thirdeye:metric:');
+
+      console.log('data', metricUrns);
+      const rows = metricUrns.map(urn => {
+        const row = {
+          urn,
+          label: toMetricLabel(urn, entities),
+          isSelected: selectedUrns.has(urn),
+          links: links[urn]
+        };
+
+        buckets.forEach((t, i) => {
+          const change = changes[urn][i];
+          row[`${t}`] = {
+            change: humanizeChange(change),
+            direction: toColorDirection(change, isInverse(urn, entities))
+          };
+          row[`${t}_raw`] = makeSortable(change);
+        });
+
+        return row;
+      });
+
+      console.log('data done', rows);
+
+      return rows;
+    }
+  ),
+
+  /**
+   * A mapping of each metric and its url(s)
+   * @type {Object} - key is metric urn, and value is an array of objects, each object has a key of the url label,
+   * and value as the url
+   * @example
+   * {
+   *  thirdeye:metric:12345: [],
+   *  thirdeye:metric:23456: [
+   *    {urlLabel: url},
+   *    {urlLabel: url}
+   *  ]
+   * }
+   */
+  links: computed('entities', function() {
+    const { entities } = this.getProperties('entities');
+    let metricUrlMapping = {};
+
+    filterPrefix(Object.keys(entities), 'thirdeye:metric:')
+      .forEach(urn => {
+        const attributes = entities[urn].attributes;
+        const { externalUrls = [] } = attributes;
+        let urlArr = [];
+
+        // Add the list of urls for each url type
+        externalUrls.forEach(urlLabel => {
+          urlArr.push({
+            [urlLabel]: attributes[urlLabel][0] // each type should only have 1 url
+          });
+        });
+
+        // Map all the url lists to a metric urn
+        metricUrlMapping[urn] = urlArr;
+      });
+
+    return metricUrlMapping;
+  }),
+
+  /**
+   * Keeps track of items that are selected in the table
+   * @type {Array}
+   */
+  preselectedItems: computed(
```

Comment:
you might be able to use `computed.filter` here.
https://www.emberjs.com/api/ember/2.15/namespaces/Ember.computed/methods/filter?anchor=filter

Something like,
preselectedItems: computed.filter( 'data', 'selectedUrns', function(urn, index, array) {
   return (urn => data[urn]).map(urn => data[urn]);
}),

Reply:
yeah. this one pulls in data from two properties. not sure if computed.filter() was designed for that. at least I can't quite find examples